providers/adfs.go

```diff
@@ -0,0 +1,191 @@
+package providers
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/requests"
+)
+
+// AdfsProvider represents an ADFS based Identity Provider
+type AdfsProvider struct {
+	*ProviderData
```

Comment:
So I want to try doing struct composition on the `OIDCProvider` for all OIDC based Providers moving forward.

Glancing over your PR and skimming the ADFS docs, ADFS appears to use OIDC instead of vanilla OAuth2?

If that's the case, lets try to use `OIDCProvider` as a base -- it gets by far the most development attention, so if you base your provider off of it, you will get feature improvements more readily.

I imagine you'd just need to override `EnrichSession` to handle the custom ADFS Userinfo endpoint & maybe `GetLoginURL`.

It looks like `ProtectedResource` might force you to have a custom `GetLoginURL` & `Redeem` though without the convenience of getting to use `c := oauth2.Config` + `c.Exchange(ctx, code)` if `ProtectedResource` needs to be a POST param 🤔

Reply:
Looking at the docs, it looks like you can potentially pass the resource parameter as part of the `scope` parameter:

>AD FS identifies the resource which the client wants to access through the resource parameter passed in the auth request. If using MSAL client library, then resource parameter is not sent. Instead the resource url is sent as a part of the scope parameter: scope = [resource url]/[scope values e.g., openid].

If we do that, then it looks like we can use the OIDCProvider as the base for nearly all functionality

Reply:
Didn't know that. I just tested it and it seems is working fine if I set the parameters:
provider: oidc
scope: https://MYRESOURCEURL/openid
user-id-claim: upn

I'll do some more tests but if everything is working as expected I guess this provider and consequently this PR is not needed at all.

Reply:
`user-id-claim` was a poorly named config flag (it affected the `session.Email` field). It has been deprecated in favor of `email-claim` if I recall.

There's nothing to override the `session.User`, it always comes from the IDToken subject.

So if you want to change that functionality, I think you would need a custom ADFS provider still that inherits most functionality from the OIDCProvider.

Out of curiosity, what do the `sub` vs `upn` claims look like with ADFS? Is replacing `sub` with `upn` the right design decision?

Reply:
> So I want to try doing struct composition on the `OIDCProvider` for all OIDC based Providers moving forward.

Do you have any example that I can follow to do that?

> Out of curiosity, what do the sub vs upn claims look like with ADFS? Is replacing sub with upn the right design decision?

According to ADFS docs (https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#using-claims-to-reliably-identify-a-user-subject-and-object-id)
"The sub claim in the Microsoft identity platform is "pair-wise" - it is unique based on a combination of the token recipient, tenant, and user. Thus, two apps that request ID tokens for a given user will receive different sub claims, but the same oid claims for that user"
So strictly speaking we shouldn't replace sub with upn. Is just that upn has a human readable value.
e.g:
"sub": "SwJXNVKWUojbGj2BploxwpEredscitQ3ScOF+N6iLCU=",
"upn": "sami.racho@mail.com",

Reply:
Unfortunately, I don't have a sample. I want to move all the OIDC based providers to be based on the central `OIDCProvider` and not `ProviderData` (which is pure OAuth2) so they stop duplicating identical OIDC code.

None of the candidates (Azure, LoginGov, Google, Keycloak, Gitlab) have had a community PR yet, so eventually I might give it a whirl myself with Keycloak.

Reply:
Oki, if you don't mind i'll try to do keep updating this pr  and try to do it by myself.

Reply:
Done some refactor using OIDCProvider as base.